README.md
# HPI Schul-Cloud Client  
_An implementation of the HPI Schul-Cloud client with NodeJS and Express._  

Dev: [![Build Status](https://travis-ci.com/hpi-schul-cloud/schulcloud-client.svg?branch=develop)](https://travis-ci.com/hpi-schul-cloud/schulcloud-client)
Master: [![Build Status](https://travis-ci.com/hpi-schul-cloud/schulcloud-client.svg?branch=master)](https://travis-ci.com/hpi-schul-cloud/schulcloud-client)

## Powered By

[![Mergify Status](https://gh.mergify.io/badges/hpi-schul-cloud/nuxt-client.png?style=cut)](https://mergify.io)
<a href="https://lokalise.com/" ><img height="18px" src="https://lokalise.com/img/lokalise_logo_black.png" style="padding: 2px 8px; border: 1px solid lightgrey; border-radius: 4px;" alt="Lokalise Logo"></a>

> # Deprecation Warning
> Please note that this client is going to be deprecated.
> All new frontend works is done in Vue.js and can be found here: https://github.com/hpi-schul-cloud/nuxt-client

## Requirements  
  
* node.js 12 or later

You might take a look at the [Dockerfile](https://github.com/hpi-schul-cloud/schulcloud-client/blob/master/Dockerfile) to see some more dependencies and latest version informations.

## Setup  
  
German docs on installing: [Setup SC](https://docs.hpi-schul-cloud.org/display/SCDOK/Setup)
  
Short version:

1. Clone directory into local folder  
2. Go into the cloned folder and enter `npm install`  
3. Install nodemon and gulp globally by entering `npm install -g nodemon gulp`  
  
## Run  
  
1. Start the [schul-cloud server](https://github.com/hpi-schul-cloud/schulcloud-server)  
2. Go into project folder
3. Run `gulp watch` to run gulp
4. Set the ENV-variable `TZ=Europe/Berlin` (for windows use `set TZ=Europe/Berlin`). You can also set the variables `SC_TITLE=HPI Schul-Cloud` and `SC_SHORT_TITLE=HPI Schul-Cloud` if you want.
5. run `npm run watch` to boot the application
6. go to `http://localhost:3100`

**Alternative with browser-sync**

1. run `gulp watch-reload` to run gulp with browser sync. It also starts the node-client-server.
2. go to `http://localhost:7000`
  
For connecting to the [SchulCloud Calendar-Service](https://github.com/hpi-schul-cloud/schulcloud-calendar) you have to set `export CALENDAR_SERVICE_ENABLED=true`.  
  
For connecting to the [SchulCloud Notification-Service](https://github.com/hpi-schul-cloud/node-notification-service) you have to set `export NOTIFICATION_SERVICE_ENABLED=true`.  
   
## Theming  
  
Add Themes to /theme directory. Call gulp and node with SC_THEME set to name of directory.  
then clear build files and gulp cache with `gulp clear`  
  
### Windows  
  run `set SC_THEME={themeName}` without spaces around the equal sign!

## How to name your branch and create a pull request (PR)
  
1. Take the Ticket Number from JIRA (ticketsystem.hpi-schul-cloud.org), e.g. SC-999  
2. Name the feature branch beginning with Ticket Number, all words separated by dash "-", e.g. `feature/SC-999-fantasy-problem`
3. Create a PR on branch develop containing the Ticket Number in PR title
4. Keep the `WIP` label as long as this PR is in development, complete PR checklist (is automatically added), keep or increase code test coverage, and pass all tests before you remove the `WIP` label. Reviewers will be added automatically. For more information check our Definition of Done [here](https://docs.schul-cloud.org/pages/viewpage.action?pageId=92831762).

## Testing  
  
**Information**: Please make sure that all your changes works on [Chrome](https://www.google.de/chrome/browser/desktop/index.html) , [Firefox](https://www.mozilla.org/de/firefox/new/) and [Safari](https://www.apple.com/de/safari/)!  
For html/css components please check [caniuse](https://caniuse.com/).  
  
1. Set the password for the demo user `schueler@schul-cloud.org`  
  Ubuntu/Mac: `export SC_DEMO_USER_PASSWORD={PASSWORD}` (Without braces)  
    Windows: `set SC_DEMO_USER_PASSWORD={PASSWORD}` (Without braces)  
2. run `npm run test`  
3. If you want to use another backend url than localhost, set the `BACKEND_URL` and `PUBLIC_BACKEND_URL` environment variables (see 1)  
4. If you want to list the coverage, run `npm run coverage`  

## Commiting

Default branch: develop

1. Go into project folder
2. Checkout to develop branch (or clone for the first time)
3. Run `git pull`
4. Create a branch for your new feature named feature/SC-*Ticket-ID*-*Description*
5. Run the tests (see above)
6. Commit with a meanigful commit message(!) even at 4 a.m. and not stuff like "dfsdfsf"
7. Start a pull request (see above) to branch develop to merge your changes

## Code rules

[CSP rules](https://github.com/hpi-schul-cloud/schulcloud-client/tree/develop/docs/CSP_RULES_DEV.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Allowed Types of change: `Added`, `Changed`, `Deprecated`, `Removed`, `Fixed`, `Security`

## [Unreleased]

### Added

### Changed

## [26.20.0] - 2021-12-7

### Changed

- BC-308 - adjust footer
- BC-617 - changes for transition to Dataport

### Fixed
- BC-856 – fixed plus sign in the add-btn

### Changed

- BC-609 - changed wording from "archive" to "finish"
- BC-308 - added data-testId to fix e2e test
- BC-741 - Sort and split ansible rols

## [26.19.0] - 2021-11-30

### Changed

- BC-661 - Introduced maximum height for school logo in top bar
- BC-740 - BC-848 - change ansible vars for docker image and tag

### Fixed

## [26.18.0] - 2021-11-26

### Fixed

-   BC-700 - fix deletion for files in tasks

### Removed

-   BC-832 - remove github image hosting from CORS

## [26.17.0] - 2021-11-16

### Changed

-   BC-589 - BC-651 - change auto deployment to a reusable workflow

### Changed

-   BC597 - update mathjax library

## [26.16.0] - 2021-11-09

### Fixed

-   BC-650 - fix header error on copy course, which was making client crush

### Changed

-   BC-404 - changed error message for team invitations via email

### Fixed

-   BC-563 - fixing link to language file for successful team permission save

## [26.15.1] - 2021-11-05

-   BC-650 - don't exit on error

## [26.15.0] - 2021-11-02

### Changed

-   BC-402 - change e2e test include to reusable workflows
-   BC-392 - adjusted skip links wording
-   BC-383 - remove usage of npm run seed in server
-   BC-525 - link faq, hover state submit-btn, btn-text Download files of ungraded task, wording homework into task
-   BC-469 - defined docker base image
-   Set data-testids in teams page for e2e purporses

### Fixed

-   BC-525 - hover state submit-btn

-   BC-379 - updated the engine version to node 16 and npm 8

## [26.14.0] - 2021-10-26

### Changed

-   BC-383 - remove usage of npm run seed in server

### Fixed

-   BC-525 - fixed footer, blog-headline landing page, logo and wording failed login

## [26.13.1] - 2021-10-22

### Changed

-   BC-379 - hard-code node 14 and npm 6

-   BC-40 - rocketchat DNS prefix adjustments for autodeployment

## [26.13.0] - 2021-10-19

### Fixed

-   BC-393 - cleaned up some translations
-   BC-405 - update dashboard pictures and replace timetable placeholder

## [26.12.0] - 2021-10-12

### Added

-   BC-343 - run unstable e2e tests on label `run unstable tests`

### Changed

-   BC-307 - design and content transformation HPI Schul-Cloud in dBildungscloud
-   BC-170 - add v1 prefix in wopi api uri extending PUBLIC_BACKEND_URL
-   BC-217 - update LibreOffic domain at http-headers.js on ansible client-configmap-files.yml.j2
-   BC-331 - BC-354 - rename ansible variables for OnePassword

### Fixed

-   BC-381 - fixed color of menu items in the sidebar
-   BC-381 - fixed colored adress-bar of browsers and added correct app-icons

## [26.11.1] - 2021-09-22

### Changed

-   BC-235 - exchange old course colors into accessible colors

## [26.11.0] - 2021-09-22

### Changed

-   BC-229 - exchange private task into draft
-   BC-94 - clean up the colorsystem and fix some UI issues
-   BC-94 - finxing margin issue sidebar current task
-   BC-329 - New school admin page button is disabled

### Fixed

-   BC-68 - fix e2e manifest missing
-   BC-242 - resetting icon for topic visibility and showing error message when permission is missing

## [26.10.2] - 2021-09-14

### Fixed

-   SC-9320 - Fix merge conflicts that throw an error in firstLogin

## [26.10.1] - 2021-09-03 --> [26.9.2]

## [26.10.0] - 2021-09-03

### Changed

-   BC-144 - Added Link to new school admin page
-   BC-44 - add api version for jwt timer service call
-   BC-5 - renamed sidebar item "Open Tasks" to "Current Tasks"
-   BC-14 - changed thr contact address
-   BC-64 - execute e2e tests via push on main
-   BC-37 - BC-55 - reduce resource consumption for deployed client
-   BC-8 - update dependencies
-   BC-81 - remove old Lern-Store

### Added

-   BC-38 - BC-125 - Add ansible files for Bosscloud (default)

## [26.9.2] - 2021-08-27

### Changed

-   BC-137 - fix bug with display of Blue Bar

## [26.9.3] - 2021-09-10

### Fixed

-   BC-239 - normalize students birth date

## [26.9.2] - 2021-08-27

### Changed

-   BC-137 - fix bug with display of Blue Bar

## [26.9.0] - 2021-08-13

### Changed

-   SC-8165 - showing a loading spinner during request for toggeling visibility of a topic
-   SC-9191 - run e2e tests in pull requests with 'run tests' label
-   SC-9192 - access feathers and nestjs apps over their dedicated version prefixes

## [26.8.1] - 2021-08-11

-   revert dependencies update (client crushed after production deployment)

## [26.8.0] - 2021-08-05

-   SC-9209 - fixed sharing course/topic text
-   SC-9074 - wording change on edit topic page
-   SC-6071 - wording change on courses page
-   SC-9236 - remove HPI footer logo from federal states instances

## [26.7.0] - 2021-07-28

### Changed

-   SC-9221 - changed contact data in imprint
-   SC-9216 - fixed Show/Hide password feature in login page
-   SC-9225 - remove legacy impressum and refer to the Nuxt impressum

### Fixed

-   SC-9212 - fix CSV import/error message formatting

## [26.6.2] - 2021-07-23

### Changed

-   SC-9190 - publish news target names

## [26.6.1] - 2021-07-21

### Changed

-   change default lernstore mode to edusharing

## [26.6.0] - 2021-07-20

### Added

-   SC-9148 - add env var GLOBAL_ANNOUNCEMENT_ADMIN_TEACHER
-   OPS-2574 - Removeing autodeployed branches for developers if branch deleted

### Removed

-   SC-8440 - remove analogue consent button from admin classes page and fix v2
-   SC-8073 - removing functionality of adding student from previous years

### Changed

-   SC-9196 - fixed submissions overview

gulpfile.js
/* eslint-disable no-console */

const autoprefixer = require('autoprefixer');
const fs = require('fs');
const gulp = require('gulp');
const babel = require('gulp-babel');
const cleanCSS = require('gulp-clean-css');
const concat = require('gulp-concat');
const gulpCount = require('gulp-count');
const gulpErrorHandler = require('gulp-error-handle');
const filelog = require('gulp-filelog');
const header = require('gulp-header');
const gulpif = require('gulp-if');
const optimizejs = require('gulp-optimize-js');
const plumber = require('gulp-plumber');
const postcss = require('gulp-postcss');
const cssvariables = require('postcss-css-variables');
const merge = require('merge-stream');
const rename = require('gulp-rename');
const rimraf = require('gulp-rimraf');
const sass = require('gulp-sass')(require('sass'));
const sassGrapher = require('gulp-sass-grapher');
const sourcemaps = require('gulp-sourcemaps');
const uglify = require('gulp-uglify');
const path = require('path');
const named = require('vinyl-named');
const webpack = require('webpack');
const webpackStream = require('webpack-stream');
const nodemon = require('gulp-nodemon');
const browserSync = require('browser-sync');
const change = require('gulp-change');
const { rewriteStaticAssetPaths } = require('./middleware/assets');
const webpackConfig = require('./webpack.config');

const browserlist = ['> 0.2%', 'last 10 version', 'not dead'];

const baseScripts = [
	'./node_modules/jquery/dist/jquery.min.js',
	'./node_modules/form-serializer/dist/jquery.serialize-object.min.js',
	'./static/scripts/tether/tether.min.js',
	'./static/scripts/bootstrap/bootstrap.min.js',
	'./static/scripts/chosen/chosen.jquery.min.js',
	'./static/scripts/base.js',
	'./static/scripts/toggle/bootstrap-toggle.min.js',
	'./static/scripts/qrcode/kjua-0.1.1.min.js',
	'./static/scripts/ajaxconfig.js',
];

// specify css files (e.g. in node modules) that should be copied to the build directory
const baseStyles = [
	{ dirname: 'calendar/', filename: 'fullcalendar.min.css', src: './node_modules/@fullcalendar/core/main.min.css' },
	{ dirname: 'calendar/', filename: 'daygrid.min.css', src: './node_modules/@fullcalendar/daygrid/main.min.css' },
	{ dirname: 'calendar/', filename: 'timegrid.min.css', src: './node_modules/@fullcalendar/timegrid/main.min.css' },
];

function themeName() {
	return process.env.SC_THEME || 'default';
}

const EXIT_ON_ERROR = process.env.GULP_EXIT_ON_ERROR
	? process.env.GULP_EXIT_ON_ERROR === 'true'
	: process.env.NODE_ENV !== 'development';

const nonBaseScripts = [
	'./static/scripts/**/*.js',
].concat(baseScripts.map((script) => `!${script}`));
// used by almost all gulp tasks instead of gulp.src(...)
// plumber prevents pipes from stopping when errors occur
// changed only passes on files that were modified since last time
// filelog logs and counts all processed files

function withTheme(src) {
	if (typeof src === 'string') {
		return [src, `./theme/${themeName()}/${src.slice(2)}`];
	}
	return src.concat(src
		.map((e) => `./theme/${themeName()}/${e.slice(2)}`));
}

const handleError = (error) => {
	console.error(error);
	process.exit(1);
};

const beginPipe = (src) => gulp
	.src(withTheme(src), { allowEmpty: true, since: gulp.lastRun('build-all') })
	.pipe(gulpif(EXIT_ON_ERROR, gulpErrorHandler(handleError), plumber()))
	.pipe(filelog());

const beginPipeAll = (src) => gulp
	.src(withTheme(src), { allowEmpty: true, since: gulp.lastRun('build-all') })
	.pipe(gulpif(EXIT_ON_ERROR, gulpErrorHandler(handleError), plumber()))
	.pipe(filelog());

// copy images
// uses gulp.src instead of beginPipe for performance reasons (logging is slow)
gulp.task('images', () => gulp
	.src(withTheme('./static/images/**/*.*'))
	.pipe(gulp.dest(`./build/${themeName()}/images`)));

// minify static/other
// uses gulp.src instead of beginPipe for performance reasons (logging is slow)
gulp.task('other', () => gulp
	.src(withTheme('./static/other/**/*.*'))
	.pipe(gulp.dest(`./build/${themeName()}/other`)));

// minify static/other
// uses gulp.src instead of beginPipe for performance reasons (logging is slow)
gulp.task('other-with-theme', gulp.series('other', () => gulp
	.src(withTheme('./static/other/**/*.*'))
	.pipe(gulp.dest(`./build/${themeName()}/other`))));

const loadPaths = path.resolve('./static/styles/');
sassGrapher.init('./static/styles/', {
	loadPaths,
});
let firstRun = true;
gulp.task('styles', () => {
	const themeFile = `./theme/${themeName()}/style.scss`;
	return beginPipe('./static/styles/**/*.{css,sass,scss}')
		.pipe(gulpif(!firstRun, sassGrapher.ancestors()))
		.pipe(header(fs.readFileSync(themeFile, 'utf8')))
		.pipe(filelog('PROCESS: '))
		.pipe(sourcemaps.init())
		.pipe(sass({
			sourceMap: true,
			includePaths: ['node_modules'],
		}).on('error', handleError))
		.pipe(postcss([
			cssvariables({
				preserve: true,
			}),
			autoprefixer({
				browsers: browserlist,
			}),
		]))
		.pipe(cleanCSS({
			compatibility: 'ie9',
		}))
		.pipe(change(rewriteStaticAssetPaths))
		.pipe(sourcemaps.write('./sourcemaps'))
		.pipe(gulp.dest(`./build/${themeName()}/styles`))
		.pipe(browserSync.stream());
});

const copyStyle = (dirname, filename, src) => gulp.src(src)
	.pipe(rename((targetPath) => {
		targetPath.basename = path.parse(filename).name;
		targetPath.dirname = dirname;
	}))
	.pipe(gulp.dest(`./build/${themeName()}/styles`));

gulp.task('copy-styles',
	() => merge(baseStyles.map(({ dirname, filename, src }) => copyStyle(dirname, filename, src))));

gulp.task('styles-done', gulp.series('styles'), () => {
	firstRun = false;
});

// copy fonts
gulp.task('fonts', () => beginPipe('./static/fonts/**/*.{eot,svg,ttf,woff,woff2}')
	.pipe(gulp.dest(`./build/${themeName()}/fonts`)));

// copy static assets
gulp.task('static', () => beginPipe('./static/*')
	.pipe(gulp.dest(`./build/${themeName()}/`)));

// compile/transpile JSX and ES6 to ES5 and minify scripts
gulp.task('scripts', () => beginPipeAll(nonBaseScripts)
	.pipe(
		named((file) => {
			// As a preparation for webpack stream: Transform nonBaseScripts paths
			// e.g. '/static/scripts/schics/schicEdit.blub.min.js' -> 'schics/schicEdit.blub.min'
			const initialPath = file.history[0].split('scripts')[1];
			const pathSegments = initialPath.split('.');
			const concretePath = pathSegments
				.slice(0, pathSegments.length - 1)
				.join('.');
			const fileName = concretePath
				.split('')
				.slice(1)
				.join('');

			return fileName;
		}),
	)
	.pipe(webpackStream(webpackConfig, webpack))
	.pipe(gulp.dest(`./build/${themeName()}/scripts`))
	.pipe(browserSync.stream()));

// compile/transpile JSX and ES6 to ES5, minify and concatenate base scripts into all.js
gulp.task('base-scripts', () => beginPipeAll(baseScripts)
	.pipe(gulpCount('## js-files selected'))
	.pipe(babel({
		presets: [
			[
				'@babel/preset-env',
				{
					modules: false,
					targets: browserlist.join(', '),
				},
			],
		],
	}))
	.pipe(optimizejs())
	.pipe(uglify())
	.pipe(concat('all.js'))
	.pipe(gulp.dest(`./build/${themeName()}/scripts`)));

// compile vendor SASS/SCSS to CSS and minify it
gulp.task('vendor-styles', () => beginPipe('./static/vendor/**/*.{sass,scss}')
	.pipe(sourcemaps.init())
	.pipe(sass({
		sourceMap: true,
	}))
	.pipe(postcss([
		autoprefixer({
			browsers: browserlist,
		}),
	]))
	.pipe(cleanCSS({
		compatibility: 'ie9',
	}))
	.pipe(sourcemaps.write('./sourcemaps'))
	.pipe(gulp.dest(`./build/${themeName()}/vendor`))
	.pipe(browserSync.stream()));

// compile/transpile vendor JSX and ES6 to ES5 and minify scripts
gulp.task('vendor-scripts', () => beginPipe('./static/vendor/**/*.js')
	.pipe(babel({
		compact: false,
		presets: [
			[
				'@babel/preset-env',
				{
					modules: false,
					targets: browserlist.join(', '),
				},
			],
		],
		plugins: ['@babel/plugin-transform-react-jsx'],
	}))
	.pipe(optimizejs())
	.pipe(uglify())
	.pipe(gulp.dest(`./build/${themeName()}/vendor`)));

// copy other vendor files
gulp.task('vendor-assets', () => beginPipe([
	'./static/vendor/**/*.*',
	'!./static/vendor/**/*.js',
	'!./static/vendor/**/*.{sass,scss}',
]).pipe(gulp.dest(`./build/${themeName()}/vendor`)));

// copy node modules
const nodeModules = {
	// example
	// 'module/path/to/keep': [
	// 	 '**/*', // matched files, e.g. copy all files in folder
	//	 'folder/**/*', // folders defined by name will be flattened
	// ],

	// mathjax
	mathjax: ['MathJax.js'],
	'mathjax/config': ['**/*'],
	'mathjax/extensions': ['**/*'],
	'mathjax/fonts': ['**/*'],
	'mathjax/jax': ['**/*'],
	'mathjax/localization': ['**/*'],

	// font-awesome
	'font-awesome/fonts': [
		'**/*',
	],

	// video.js
	'video.js/dist': ['video.min.js'],
	'video.js/dist/lang': ['*.js'],
};
gulp.task('node-modules', () => {
	const promises = [];

	for (const [module, modulePaths] of Object.entries(nodeModules)) {
		promises.push(
			gulp.src(modulePaths.map((modulePath) => `./node_modules/${module}/${modulePath}`))
				.pipe(gulp.dest(`./build/${themeName()}/vendor-optimized/${module}`)),
		);
	}

	return Promise.all(promises);
});

// clear build folder + smart cache
gulp.task('clear', () => gulp
	.src(
		[
			'./build/*',
			'./.gulp-changed-smart.json',
			'./.webpack-changed-plugin-cache/*',
		],
		{
			read: false,
			allowEmpty: true,
		},
	)
	.pipe(rimraf()));

// clear gulp cache without removing current build
gulp.task('clear-cache', () => gulp
	.src(
		[
			'./.gulp-changed-smart.json',
			'./.webpack-changed-plugin-cache/*',
		],
		{
			read: false,
			allowEmpty: true,
		},
	)
	.pipe(rimraf({})));

// run all tasks, processing changed files
gulp.task('build-all', gulp.series(
	'images',
	'other',
	'fonts',
	'other-with-theme',
	'node-modules',
	'styles',
	'styles-done',
	'copy-styles',
	'scripts',
	'base-scripts',
	'vendor-styles',
	'vendor-scripts',
	'vendor-assets',
	'static',
));

gulp.task('build-theme-files', gulp.series('styles', 'styles-done', 'images', 'static'));

// watch and run corresponding task on change, process changed files only
gulp.task('watch', gulp.series('build-all', () => {
	const watchOptions = { interval: 1000 };
	gulp.watch(baseScripts, watchOptions, gulp.series('base-scripts'));
	gulp.watch(
		withTheme('./static/styles/**/*.{css,sass,scss}'),
		watchOptions,
		gulp.series('styles', 'styles-done'),
	);
	gulp.watch(withTheme('./static/images/**/*.*'), watchOptions, gulp.series('images'))
		.on('change', browserSync.reload);
	gulp.watch(withTheme(nonBaseScripts), watchOptions, gulp.series('scripts'));

	gulp.watch(withTheme('./static/vendor/**/*.*'), watchOptions, gulp.series('vendor-styles',
		'vendor-scripts',
		'vendor-assets'));
	gulp.watch(withTheme('./static/*.*'), watchOptions, gulp.series('static'));
}));

gulp.task('nodemon', (cb) => {
	let started = false;
	return nodemon({
		ext: 'js hbs json',
		script: './bin/www',
		watch: ['views/', 'controllers/', 'helpers'],
		exec: 'node --inspect=9310',
	}).on('start', () => {
		if (!started) {
			cb();
			started = true;
		}
		setTimeout(browserSync.reload, 3000); // server-start takes some time
	});
});

gulp.task('browser-sync', () => {
	browserSync.init(null, {
		proxy: 'http://localhost:3100',
		open: false,
		port: 7000,
		ghostMode: false,
		reloadOnRestart: false,
		socket: {
			clients: {
				heartbeatTimeout: 60000,
			},
		},
	});
});

gulp.task('watch-reload', gulp.parallel('watch', 'nodemon', 'browser-sync'));


// run this if only 'gulp' is run on the commandline with no task specified
gulp.task('default', gulp.series('build-all'));
